client_cli/src/main.rs

```diff
@@ -111,27 +111,55 @@ pub enum Subcommand {
     Json(json::Args),
 }
 
+/// Context inside which command is executed
+pub trait RunContext {
+    /// Get access to configuration
+    fn configuration(&self) -> &ClientConfiguration;
+    /// Serialize and print data
+    ///
+    /// # Errors
+    /// - if serialization fails
+    /// - if printing fails
+    fn print_data(&mut self, data: &dyn Serialize) -> Result<()>;
+}
+
+struct PrintJsonContext<W> {
+    write: W,
+    config: ClientConfiguration,
+}
+
+impl<W: std::io::Write> RunContext for PrintJsonContext<W> {
+    fn configuration(&self) -> &ClientConfiguration {
+        &self.config
+    }
+
+    fn print_data(&mut self, data: &dyn Serialize) -> Result<()> {
+        writeln!(&mut self.write, "{}", serde_json::to_string_pretty(data)?)?;
+        Ok(())
+    }
+}
+
 /// Runs subcommand
 pub trait RunArgs {
     /// Runs command
     ///
     /// # Errors
     /// if inner command errors
-    fn run(self, cfg: &ClientConfiguration) -> Result<Box<dyn Serialize>>;
+    fn run(self, context: &mut dyn RunContext) -> Result<()>;
```

Comment:
Same as with `dyn Serialize`: why not use just generics?

```suggestion
    fn run<C: RunContext>(self, context: &mut C) -> Result<()>;
```

Reply:
Answered in the comment above.